### PR TITLE
mon: avoid daemons crashing with mal-formatted cmd input

### PR DESCRIPTION
**Pull request by gaord**
From a complete code analysis,code under src/common, src/messages, src/mon and src/osd get reviewed to make sure all cmd_getval stuff are running safe with these commits.
in fact, there are 364 calling locations altogether as below:
src/common(14)
src/messages(6)
src/mon(308)
src/osd(29)

Most of them above had been handled well with try/catch. This PR just created a full coverage with the latest code in order to strengthen ceph cluster from DOS attacks. please help review for better.

bug to fix:https://tracker.ceph.com/issues/57890
related bug: https://tracker.ceph.com/issues/54558

Signed-off-by: Ben Gao <bengao168@msn.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by gaord** [comment]
@pjjw Hi, the commit should be clean-handed with ceph windows tests failure? What do you think? Thank you.

**Review by idryomov — COMMENTED** [review]
This is up to @ceph/core but if it was me, I'd look into changing `cmd_getval` interface to get rid of exceptions instead of trying to sprinkle more try/catch around.  The current interface (bool return value + potential exception) is clearly too confusing, just judging by the number of places where it appears to be misused.

**Comment by gaord** [comment]
@idryomov @

> This is up to @ceph/core but if it was me, I'd look into changing `cmd_getval` interface to get rid of exceptions instead of trying to sprinkle more try/catch around. The current interface (bool return value + potential exception) is clearly too confusing, just judging by the number of places where it appears to be misused.

reviewing code could tell that tons of cases were already consolidated good by try/catch. This pr just included cases slipped through the net. thinking about "ceph tell" command may send Mal-formatted string to any daemons, eventually get them crashed, it's appealing to fix it soon with less impact one way or another.

File: src/messages/MMonCommand.h [hunk 1]
```diff
@@ -55,22 +59,27 @@ class MMonCommand final : public PaxosServiceMessage {
...
+        std::string key;
+        cmd_getval(cmdmap, "key", key);
+        o << "[{prefix=" << prefix << ", key=" << key << "}]";
+      } else {
+        for (unsigned i=0; i<cmd.size(); i++) {
+          if (i) o << ' ';
+          o << cmd[i];
+        }
       }
+    } catch (bad_cmd_get& e) {
```
Comment (rzarzynski) [inline src/messages/MMonCommand.h:80]: OK, so most of the changes in this commit is about reindenting `cmd_getval()`-related blocks due to catching `bad_cmd_get`.

File: src/mon/Monitor.cc [hunk 2]
```diff
@@ -4534,7 +4544,12 @@ void Monitor::dispatch_op(MonOpRequestRef op)
       return;
     case MSG_COMMAND:
       op->set_type_command();
-      handle_tell_command(op);
+      try {
+        handle_tell_command(op);
+      } catch (const bad_cmd_get& e) {
+        reply_command(op, -EINVAL, e.what(), 0);
```
Comment (rzarzynski) [inline src/mon/Monitor.cc:4550]: `MMonCommandAck` the `reply_command()` generates is finally handled by `MonClient` which, as far as I see, doesn't use the version (`0` here) field.

**Review by rzarzynski — APPROVED** [review]

**Review by ljflores — CHANGES_REQUESTED** [review]
Hey @gaord can you fix the commit titles so they align with https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst#commit-title?

**Comment by NitzanMordhai** [comment]
Rados review : https://tracker.ceph.com/projects/rados/wiki/MAIN#httpspulpitocephcomyuriw-2023-10-30_153436-rados-wip-yuri10-testing-2023-10-27-0804-distro-default-smithi

**Comment by yuriw** [comment]
jenkins test windows

**Comment by yuriw** [comment]
jenkins test make check

**Comment by yuriw** [comment]
@gaord this was tested and ready for merge after checks passed and comments resolved
ref: https://trello.com/c/IdHRIj6u

**Comment by ljflores** [comment]
> Hey @gaord can you fix the commit titles so they align with https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst#commit-title?

@gaord this is the only thing missing; when this is fixed, the PR can be merged.

**Comment by gaord** [comment]
> > Hey @gaord can you fix the commit titles so they align with https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst#commit-title?
> 
> @gaord this is the only thing missing; when this is fixed, the PR can be merged.

missed this message among lots of others.  What shall the title look like? would like to make a push, not clear what to change, however. Thanks.
@ljflores

**Pull request closed without merging**